Add support for automatic file nesting

rodrmoya opened the pull request:
Fixes https://devdiv.visualstudio.com/DevDiv/_workitems/edit/781656
Fixes https://devdiv.visualstudio.com/DevDiv/_workitems/edit/857165

Commit:
[Core] Add (very) basic FileNestingService

Commit:
[Core] Implement basic skeleton of FileNestingService

Since we will be loading the rules from files, and even though the initial
version will just use a single one, prepare implementation for future
expansion to support more than 1 file, each of which is implemented
in a NestingRulesProvider, which in turn contains a set of NestingRule's.

Therzok commented:
This should be easily testable. Needs unit tests.

rodrmoya commented:
> This should be easily testable. Needs unit tests.

Yeah, work in progress, and will add lots of tests

Commit:
[Core] Add basic FileNestingService's tests

Commit:
[Core] Add support for ExtensionToExtension nesting rule

Commit:
[Core] Add support for FileSuffixToExtension nesting rule

Commit:
[Core] Add support for PathSegment nesting rule

Commit:
[Core] Add support for AllExtensions nesting rule

Commit:
[Core] Take into account extensions and patterns for PathSegment rule

Commit:
[Core] Add support for FileToFile nesting rule

Commit:
[Core] Add support for reading nesting rules from files

mrward commented on main/src/core/MonoDevelop.Core/MonoDevelop.Projects.FileNesting/FileNestingService.cs:
```diff
@@ -0,0 +1,55 @@
+namespace MonoDevelop.Projects.FileNesting
+{
+	public static class FileNestingService
+	{
```
Does this need to be public? MonoDevelop.Core.Tests should have access to MonoDevelop.Core's internals.

rodrmoya replied:
I was planning to inject the .json from the AspNetCore extension, so that each extension can provide the nesting rules they want, so that's why I made it public. But maybe there's a better way to inject it? Or maybe we should just have the asp.net core rules available to all projects?

rodrmoya replied:
Moved it back to not public, as I implemented it via ExtensionPoint/Extension in https://github.com/mono/monodevelop/pull/7682/commits/627693a8d52f0768f716f1132173015d701f0b50

mrward commented on main/src/core/MonoDevelop.Core/MonoDevelop.Projects.FileNesting/NestingRule.cs:
```diff
@@ -0,0 +1,146 @@
+				if (inputFile.EndsWith (AppliesTo, StringComparison.OrdinalIgnoreCase)) {
+					foreach (var pt in patterns) {
+						parentFile = inputFile.Replace (AppliesTo, pt);
+						if (CheckParentForFile (inputFile, parentFile)) {
```
This looks like it will replace all occurrences of AppliesTo in the file path? I think we want to only replace the last part.

rodrmoya replied:
Right

rodrmoya replied:
Fixed now

mrward commented on main/src/core/MonoDevelop.Core/MonoDevelop.Projects.FileNesting/NestingRule.cs:
```diff
@@ -0,0 +1,146 @@
+		}
+
+		public string GetParentFile (string inputFile)
+		{
```
Just wondering if there is a reason to not use FilePath here for the inputFile and the return value? I do not have a preference either way, but when I see `Path.` methods being used I usually think about using the FilePath class.

rodrmoya replied:
Done, will push shortly

Commit:
[Core] Fix parsing of arrays in the JSON

Commit:
[Core] Add missing return that was making tests fail

Commit:
[Core] Replace only last occurrence of the pattern for FileSuffixToExtension rules

Commit:
[AspNetCore] Implement a NestingRulesProvider for ASP.NET Core

Commit:
[Core] Extend FileNestingService test

Test also the case where the parent file does not exist, in which case
the FileNestingService shouldn't return anything.

Commit:
[AspNetCore] Check file nesting rules for ASP.NET Core project tests

Commit:
[Core] Add more console output for the failing test

Therzok commented on main/src/addins/MonoDevelop.AspNetCore/MonoDevelop.AspNetCore/AspNetCoreNestingRulesProvider.cs:
```diff
@@ -0,0 +1,40 @@
+namespace MonoDevelop.AspNetCore
+{
+	[Extension]
+	class AspNetCoreNestingRulesProvider : NestingRulesProvider
```
```suggestion
	sealed class AspNetCoreNestingRulesProvider : NestingRulesProvider
```

Therzok commented on main/src/addins/MonoDevelop.AspNetCore/MonoDevelop.AspNetCore/AspNetCoreNestingRulesProvider.cs:
```diff
@@ -0,0 +1,40 @@
+	{
+		public AspNetCoreNestingRulesProvider ()
+		{
+			SourceFile = AddinManager.CurrentAddin.GetFilePath ("Resources/AspNetCore.filenesting.json");
```
```suggestion
			SourceFile = AddinManager.CurrentAddin.GetFilePath (Path.Combine ("Resources", "AspNetCore.filenesting.json");
```

Therzok commented on main/src/core/MonoDevelop.Core/MonoDevelop.Projects.FileNesting/FileNestingService.cs:
```diff
@@ -0,0 +1,59 @@
+		public static string GetParentFile (string inputFile)
+		{
+			if (rulesProviders == null) {
+				rulesProviders = new List<NestingRulesProvider> ();
```
We're caching the list, but the providers can change at runtime. Listen for extension changes to get providers.

rodrmoya replied:
Yes, although right now we just want to support the ASP.NET Core rules (or, if we find it useful, the "standard" plugins), so want to keep it as simple as possible, but because of the failing test in CI (https://devdiv.visualstudio.com/DevDiv/_releaseProgress?_a=release-environment-logs&releaseId=357250&environmentId=1668000) I might have to some small refactoring to the `FileNestingService`, so might include that

Therzok replied:
Disabling/enabling the aspnetcore addin can break this code. It's a matter of reliability, not simplicity here.

rodrmoya replied:
Done

Therzok commented on main/src/core/MonoDevelop.Core/MonoDevelop.Projects.FileNesting/FileNestingService.cs:
```diff
@@ -0,0 +1,59 @@
+
+			foreach (var rp in rulesProviders) {
+				var parentFile = rp.GetParentFile (inputFile);
+				if (!String.IsNullOrEmpty (parentFile)) {
```
```suggestion
				if (!string.IsNullOrEmpty (parentFile)) {
```

rodrmoya replied:
I'm going to change this to use FilePath's, so not applying this commit, as this will change

Therzok commented on main/src/core/MonoDevelop.Core/MonoDevelop.Projects.FileNesting/NestingRule.cs:
```diff
@@ -0,0 +1,147 @@
+		public NestingRule (NestingRuleKind kind, string appliesTo, IEnumerable<string> patterns)
+		{
+			Kind = kind;
+			AppliesTo = appliesTo ?? ".*";
```
```suggestion
			AppliesTo = appliesTo ?? AllFilesWildcard;
```

Therzok replied:
Also, what happens when appliesTo is `string.Empty`?

Therzok commented on main/src/core/MonoDevelop.Core/MonoDevelop.Projects.FileNesting/NestingRule.cs:
```diff
@@ -0,0 +1,147 @@
+		{
+			Kind = kind;
+			AppliesTo = appliesTo ?? ".*";
+			this.patterns = patterns?.ToList ();
```
Patterns is used later on, and it can be null here. Possible NREs.

rodrmoya replied:
right, fixed now

Therzok commented on main/src/core/MonoDevelop.Core/MonoDevelop.Projects.FileNesting/NestingRulesProvider.cs:
```diff
@@ -0,0 +1,150 @@
+
+		public NestingRulesProvider (string fromFile)
+		{
+			SourceFile = fromFile;
```
Check argument for null?

rodrmoya replied:
👍

rodrmoya committed:
[AspNetCore] Seal AspNetCoreNestingRulesProvider class

Co-Authored-By: Marius Ungureanu <teromario@yahoo.com>

rodrmoya committed:
[AspNetCore] Use Path.Combine

Co-Authored-By: Marius Ungureanu <teromario@yahoo.com>

rodrmoya committed:
[AspNetCore] Use wildcard constant

Co-Authored-By: Marius Ungureanu <teromario@yahoo.com>

Commit:
[AspNetCore] Add missing using statement

Commit:
[Core] Update DependsOn on renames only if it was already set

If not, when renaming files that have been nested via the FileNestingService,
an unsolicited DependentUpon entry is saved to the csproj.

Commit:
[Core] Avoid a null patterns list

Commit:
[Core] Listen to extension node changes for NestingRulesProvider's

mrward commented on main/src/addins/MonoDevelop.AspNetCore/MonoDevelop.AspNetCore/AspNetCoreNestingRulesProvider.cs:
```diff
@@ -0,0 +1,41 @@
+namespace MonoDevelop.AspNetCore
+{
+	[Extension]
+	sealed class AspNetCoreNestingRulesProvider : NestingRulesProvider
```
Should we support an id here to allow ordering of file nesting providers? Not sure how to do this using the attribute - usually do this in the .addin.xml.

If I write a file nesting provider and want it to be used instead of the AspNetCoreNestingRulesProvider we can do that if each one has an id, using insertbefore/insertafter.

rodrmoya replied:
Good question. Right now, there's only 1 `NestingRulesProvider`, but indeed we might want to add more. I've been thinking about this, and maybe a solution would be to associate those providers with specific project types (or all), so the current provider would just apply to ASP.NET Core projects, another one could be added for Xamarin.ios/Mac/Android, etc.

About the `insertAfter/insertBefore`, not sure, as in theory the order should be a configurable setting for the user (once we support custom file nesting UI), but might be an option.

mrward replied:
OK - if these can be ordered by the user, which seems to be the case in VS Win, then insertbefore/after is not useful here. I guess the ordering will need to be defined somewhere else and applied in the FileNestingService.

mrward commented on main/src/core/MonoDevelop.Core/MonoDevelop.Projects.FileNesting/NestingRulesProvider.cs:
```diff
@@ -0,0 +1,150 @@
+		public string GetParentFile (string inputFile)
+		{
+			if (nestingRules == null) {
+				if (!File.Exists (SourceFile)) {
```
If the source file does not exist or fails to load should we set a flag the first time this happens instead of checking File.Exists and trying to load every time GetParentFile is called?

rodrmoya replied:
hmm, but what if the file gets created after? Not that it can happen right now, as there's only 1 provider that gets loaded on startup, but it might happen when we have more. You mean then just marking as not existing and use FSW to detect if/when it gets created?

mrward replied:
Mostly thinking about avoiding File.Exists being called for each file in the Solution pad. Was not thinking about reloading of the file due to changes by the user.

rodrmoya replied:
ah, but this `SourceFile` is the .json file with the rules, not every file in the solution. But, that said, I think I do a File.Exists for every file returned by this, so will check.

mrward commented on main/tests/MonoDevelop.Core.Tests/MonoDevelop.Projects/MSBuildGlobTests.cs:
```diff
@@ -1379,6 +1379,7 @@ public async Task RenameXamlAndXamlCSharpFileAtSameTime (bool addXamlFilesBefore
 				// Project xml should be unchanged.
 				var projectXml = File.ReadAllText (p.FileName);
+				Console.WriteLine ($"Comparing expected {expectedProjectXml} to {projectXml}");
 				Assert.AreEqual (expectedProjectXml, projectXml);
```
Debug code?

rodrmoya replied:
Removed

iantoalms commented on main/src/core/MonoDevelop.Core/MonoDevelop.Projects.FileNesting/NestingRulesProvider.cs:
```diff
@@ -0,0 +1,150 @@
+		void AddRule (NestingRuleKind kind, string appliesTo, IEnumerable<string> patterns)
+		{
+			if (nestingRules == null) {
+				nestingRules = new List<NestingRule> ();
```
Why not just use a field initializer and avoid the null check

rodrmoya replied:
The null value is used to know if the rules need to be loaded from file, so if I field initialise it, the check here (https://github.com/mono/monodevelop/pull/7682/files/8efa0b1234e601b97eb4e3a9a21e86d964d862f3#diff-9f0f0b39ec9b79ec2d5db396c3fde582R128) would just have to check for ' Count` on the collection, and in some cases it might do an extra read from the file every time it is called.

iantoalms commented on main/src/core/MonoDevelop.Core/MonoDevelop.Projects.FileNesting/NestingRulesProvider.cs:
```diff
@@ -0,0 +1,150 @@
+		static bool LoadFromFile (NestingRulesProvider provider)
+		{
+			try {
+				using (var reader = new StreamReader (provider.SourceFile)) {
```
The new syntax for this is great: https://docs.microsoft.com/en-us/dotnet/csharp/whats-new/csharp-8#using-declarations

rodrmoya replied:
hmm, something's not setup correctly for C#8 then, as changing the code to do that shows a warning, about the statement being empty and `reader` not being defined:

<img width="531" alt="image" src="https://user-images.githubusercontent.com/1873039/59854031-5b5eb900-9372-11e9-90bf-f13d7fd5ec00.png">

@Therzok ? this is vsmac 8.2 (build 832)

rodrmoya replied:
C# language version is set to 8 for that project

Therzok replied:
it should be:
```csharp
using var reader = new StreamReader (provider.SourceFile);
```

Therzok replied:
Putting the extra set of parens makes it behave as if it were:

```csharp
using (var reader = new StreamReader (provider.SourceFile)) {}
```

rodrmoya replied:
ugh, indeed 👍

iantoalms commented on main/src/core/MonoDevelop.Core/MonoDevelop.Projects.FileNesting/NestingRulesProvider.cs:
```diff
@@ -0,0 +1,150 @@
+					var json = JObject.Parse (reader.ReadToEnd ());
+					if (json != null) {
+						var parentNode = json ["dependentFileProviders"] [TokenNameAdd] as JObject;
+						foreach (var rp in parentNode.Properties ()) {
```
nit: rp? Would be nice to have a more descriptive name

rodrmoya replied:
Changed

iantoalms commented on main/src/core/MonoDevelop.Core/MonoDevelop.Projects.FileNesting/FileNestingService.cs:
```diff
@@ -0,0 +1,76 @@
+
+		public static string GetParentFile (string inputFile)
+		{
+			if (rulesProviders == null) {
```
Just field initialize rulesProviders and avoid the null check here and above?

rodrmoya replied:
Same reason as above, really want to make sure file is read only once

iantoalms commented on main/src/core/MonoDevelop.Core/MonoDevelop.Projects.FileNesting/NestingRulesProvider.cs:
```diff
@@ -0,0 +1,150 @@
+
+		public string GetParentFile (string inputFile)
+		{
+			if (nestingRules == null) {
```
This can be considerably simplified. If you get rid of the null check by having a field initializer:

		if (File.Exists (SourceFile) && LoadFromFile (this)) {
			foreach (var rule in nestingRules) {
				string parentFile = rule.GetParentFile (inputFile);
				if (!String.IsNullOrEmpty (parentFile)) {
					// Stop at the 1st rule found
					return parentFile;
				}
			}
		}
		return null;

rodrmoya replied:
But we don't want to load the file every single time, just once, hence the null initial value/check.

Commit:
[Tests] Remove debug code

Commit:
[Projects] Add null check in NestingRulesProvider's constructor

Commit:
[Projects] Use a more descriptive var name

Commit:
[Projects] Simplify ImmutableList creation for an IEnumerable

Commit:
[Projects] Use FilePath in API refering to files

Therzok commented on main/src/core/MonoDevelop.Core/MonoDevelop.Projects.FileNesting/FileNestingService.cs:
```diff
@@ -0,0 +1,73 @@
+{
+	internal static class FileNestingService
+	{
+		static ImmutableList<NestingRulesProvider> rulesProviders;
```
Why are we using an ImmutableList here?

rodrmoya replied:
It's a very small collection (only 1 item max now) and solves concurrent access, which is needed because the ExtensionNodeHandler can be ran on a separate thread than the one used for resolving, although I guess it could use a Concurrent* collection instead, although for such a small collection, not sure it's worth.

Therzok replied:
Understood, now I see the _why_.

Commit:
[Projects] Use C# 8 using syntax

jtorres reviewed (approved):
apart of the moving files issue, looks good

mrward commented on main/src/core/MonoDevelop.Core/MonoDevelop.Projects.FileNesting/NestingRulesProvider.cs:
```diff
@@ -0,0 +1,151 @@
+		public FilePath GetParentFile (FilePath inputFile)
+		{
+			if (nestingRules == null) {
+				if (!File.Exists (SourceFile)) {
```
Still doing a File.Exists here. Which will be called for every file in the project if the rules file does not exist, or if the load fails.

rodrmoya replied:
Right, fixed now

Commit:
[Projects] Avoid re-loading of files if it failed the first time

If a nesting rules file failed to load, we were re-trying reading it again
and again, for every file in the solution, so just make sure we try to load
it once only, and if it fails, just ignore it.

mrward commented on main/src/core/MonoDevelop.Core/MonoDevelop.Projects.FileNesting/FileNestingService.cs:
```diff
@@ -0,0 +1,73 @@
+			return null;
+		}
+	}
+}
```
This looks like an unusual pattern. The existing code either calls AddinManager.GetExtensionObjects<T> so the tree is re-created every time, or it uses AddinManager.AddExtensionNodeHandler. No code uses both.

If you call AddinManager.AddExtensionNodeHandler then this will be called [multiple times initially with all addins](https://github.com/mono/mono-addins/blob/a543fa0011098232041c99bc5acb6416cb4acbe0/Mono.Addins/Mono.Addins/ExtensionNode.cs#L170).

So I think we should one of these approaches. Since this is going to be called a lot I think it makes sense to catch the rule providers and handle changes using just AddinManager.AddExtensionNodeHandler.

mrward commented on main/src/core/MonoDevelop.Core/MonoDevelop.Projects.FileNesting/NestingRule.cs:
```diff
@@ -0,0 +1,148 @@
+				string inputFileString = inputFile;
+				if (inputFileString.EndsWith (AppliesTo, StringComparison.OrdinalIgnoreCase)) {
+					int suffixPosition = inputFileString.LastIndexOf (AppliesTo, StringComparison.OrdinalIgnoreCase);
+					foreach (var pt in patterns) {
```
If inputFileFile ends with AppliesTo it seems that we could find the index by substracting the length of AppliesTo from inputFileString instead of using LastIndexOf here.

rodrmoya replied:
Fixed in https://github.com/mono/monodevelop/pull/7682/commits/2f22de536e952a2dff9f67b055fe288b29776b36

mrward commented on main/src/core/MonoDevelop.Core/MonoDevelop.Projects.FileNesting/NestingRule.cs:
```diff
@@ -0,0 +1,148 @@
+		bool CheckParentForFile (FilePath inputFile, FilePath parentFile)
+		{
+			if (File.Exists (parentFile) && inputFile != parentFile) {
+				LoggingService.LogInfo ($"Applied rule for nesting {inputFile} under {parentFile}");
```
I am just wondering if there is a better way of doing this which avoids using File.Exists. Should we instead be using the Project.Files information? That would avoid using File.Exists. I am assuming that CheckParentForFile will be called after the Project.Files collection has been populated.

Do we want file nesting to apply to files that are not explicitly in the project? Not sure it has any affect in practice - unless if we enable show all files would files be nested then?

rodrmoya replied:
Yes, I think the best would be to pass the `Project` and check the `Files` collection indeed.

rodrmoya replied:
Fixed in https://github.com/mono/monodevelop/pull/7682/commits/cf69e5a72e781f80565958a5d886d8fdaef1c19a

mrward commented on main/src/core/MonoDevelop.Core/MonoDevelop.Projects.FileNesting/NestingRulesProvider.cs:
```diff
@@ -0,0 +1,156 @@
+		void AddRule (NestingRuleKind kind, string appliesTo, IEnumerable<string> patterns)
+		{
+			if (nestingRules == null) {
+				nestingRules = new List<NestingRule> ();
```
Minor: I guess this check is not needed now. nestingRules is initialized in GetParentFile now.

rodrmoya replied:
https://github.com/mono/monodevelop/pull/7682/commits/f0b97debe27c97833fe9522ec7f9205a8d365bfb

Commit:
[Projects] Remove unneded code to query all extensions

We keep the list of NestingRulesProvider's up to date already by listening
to the extension node changes.

Commit:
[Projects] Pass Project to FileNestingService

Thus, we can just check inside the project's Files collection, instead
of calling File.Exists().

Commit:
[Projects] Avoid using LastIndexOf when not necessary

Commit:
[Projects] Remove unneeded check

The nestingRules list is now created the first time GetParentFile is called,
so when this code is hit, the list should already have been created.

mrward reviewed (approved):
Code looks good to me.

Commit:
[Projects] Fix concurrent access to FileNestingService

The providers are initialized on demand, so in solutions with several
projects, that load each project on a separate Task, we were getting race
conditions on the NestingRulesProvider's initialization, so make sure
they are initialized only once and not called until they are fully loaded.

Therzok commented:
We should've added a performance test for this before merging?
